csrc/dynamic_transform.cpp

```diff
@@ -0,0 +1,477 @@
+    auto incomplete_out_tv = kv.first;
+    const auto view_analysis = kv.second;
+
+    auto inp_tv = ir_utils::producerTvsOf(incomplete_out_tv).at(0);
+
+    auto concrete_reshape_out_tv = reshape(inp_tv, view_analysis);
+
+    // Replace the old tensor with the new concretized tensor
+    for (auto use_of_old_tv : incomplete_out_tv->uses()) {
+      ir_utils::replaceValInExpr(
+          use_of_old_tv, incomplete_out_tv, concrete_reshape_out_tv);
+    }
+
+    if (incomplete_out_tv->isFusionOutput()) {
+      incomplete_out_tv->fusion()->replaceOutput(
+          incomplete_out_tv, concrete_reshape_out_tv);
+    }
+
+    incomplete_out_tv->fusion()->removeVal(incomplete_out_tv);
+  }
+}
+
+// Concretizes inherited symbolic domains. Note that when this is
+// called, it is assumed that all dynamic ops themselves are
+// concretized. Since symbolic IDs may be propagated down to
+// consumers, those domains need to be concretized accordingly.
+void DynamicTransformConcretizer::mutate(TensorView* tv) {
+  if (!tv->domain()->hasSymbolicAxis()) {
+    return;
+  }
+
+  // First, try to concretize the root domain as there may be symbolic
+  // axes inherited from the producers
+  auto propagated = propagateFromProducerToConsumer(tv);
+
+  // If no root domain is altered, nothing to do further
+  if (!propagated) {
+    return;
+  }
+
+  // Root IDs are altered. Need to propagate the changes to rfactor
+  // domain
+
+  // At this point, there should be no expr beyond rfactor root
+  TORCH_INTERNAL_ASSERT(
+      tv->domain()->domain() == tv->getMaybeRFactorDomain(),
+      "Invalid tensor: ",
+      tv->toString());
+
+  // If it has an rfactor root domain, the IterTypes of the rfactor
+  // IDs may need to be updated as well. Traverse the rfactor exprs
+  // and mutate the IterTypes of output IDs if symbolic.
+  if (tv->hasRFactor()) {
+    // Note that it is assumed that theres's no further expression
+    // beyond the rfactor domain as asserted above
+    auto all_id_stmts = StmtSort::getStmtsBetween(
+        tv->fusion(),
+        {tv->getRootDomain().begin(), tv->getRootDomain().end()},
+        {tv->getMaybeRFactorDomain().begin(),
+         tv->getMaybeRFactorDomain().end()});
+    for (auto expr : ir_utils::filterByType<Expr>(all_id_stmts)) {
+      // If none of the output IDs is symbolic, nothing to concretize
+      if (std::all_of(
+              expr->outputs().begin(), expr->outputs().end(), [](Val* output) {
+                TORCH_INTERNAL_ASSERT(output->isA<IterDomain>());
+                return output->as<IterDomain>()->getIterType() !=
+                    IterType::Symbolic;
+              })) {
+        continue;
+      }
+      // If any of output IDs is symbolic, all outputs should be symbolic
+      TORCH_INTERNAL_ASSERT(std::all_of(
+          expr->outputs().begin(), expr->outputs().end(), [](Val* output) {
+            return output->as<IterDomain>()->getIterType() ==
```

Comment:
nit: This should be safe but I'm generally in the habit of filtering IterDomains anyways (since it's not safe on inputs of transform exprs)

Reply:
Makes sense, but I actually think filtering may not be the best in this case. If there's any output that's not an `IterDomain`, we should not filter it out silently as it's likely the logic here needs to be updated, so instead of filtering I added an explicit assertion at the beginning of this loop.